### ISSUE TITLE
Code in VSCode deletes recent saved changes when connecting signal from Editor

### ISSUE THREAD

**Issue by jasoben** [godotengine/godot#91314]
### Tested versions

4.2.1 stable

### System information

OSX 14.4.1

### Issue description

When I connect a signal to a node, recent saved changes to the file on that node are deleted.

### Steps to reproduce

1. Write code in file MYFILE.gd using VSCode
2. Save code
3. Connect signal from Area2D to Node with MYFILE.gd
4. Go back to VSCode
5. Saved changes from step 2 are gone

### Minimal reproduction project (MRP)

I can't reproduce the problem as it doesn't happen all the time.

**Comment by AThousandShips** [comment]
Sounds related to:
* https://github.com/godotengine/godot/issues/88399

**Comment by PokkeFe** [comment]
This issue still occurs in v4.3.dev6.mono.official [89850d553]

External Editor Flags: `{project} --goto {file}:{line}:{col}`

**Comment by Morokiane** [comment]
If the script is open in the script editor when you connect the signal; it'll update the script in editor and overwrite the one in the external editor.

**Comment by jasoben** [comment]
Thanks @Morokiane I don't use the built-in script editor, but I wonder if there is a similar mechanism happening regardless when the signal is connected?

Regardless I will try updating to the latest stable version and keeping an eye out for the bug. 

Do any of you have suggestions on what kind of log data I could look for when the bug happens? It seems to occur so randomly it's hard to pinpoint what causes it.

My *very uneducated* imagination of the problem is that Godot is caching the script and that the bug happens when there is a hiccup in communication between VSCode and Godot. If Godot updates the cached version and tells VSCode that this is the new version, VSCode accepts this without pushback. Therefore there must be some circumstance that breaks the link between VSCode and Godot. 

If anybody more familiar with Godot's architecture could verify that idea and tell me where to find logs that might be generated (or how to set up some logging apparatus), I would be happy to do what I can to use my experience as a way to pinpoint the bug.

**Comment by Morokiane** [comment]
What I have found is Godot will open a script in the internal editor if there is a crash, even when it does bring the external editor forward with the error. At that point because the script ends up getting opened the signal connection overwriting the external version is more prone to happen.

There are cases where it external editor version will get overwritten without the signal connections, but those are rare and I haven't found the pattern for it. 

It all seems to stem from Godot opening the scripts in the internal editor on crashes.

**Comment by jasoben** [comment]
Huh, good to know. I'll look out for that.

**Comment by jasoben** [comment]
Update to this issue: 

What I found is that this has nothing to do with Signals as originally thought. The problem happens thusly:

1. I run the project from the Godot Editor (as opposed to Run from inside VSCode)
2. There is an error
3. The Godot Editor opens the script for debugging (but doesn't switch to the script window automatically)
4. The script remains open in the background
5. Any changes I make to the script in VSCode are overwritten the next time I run the game from the Godot Editor b/c it is giving preference to the open script in the Godot Script Editor

How I solved it:

Script Editor > Debug Menu > Debug With External Editor

Once I checked this box I no longer had the problem.

**Comment by christian-reichart** [comment]
> Update to this issue:
> 
> What I found is that this has nothing to do with Signals as originally thought. The problem happens thusly:
> 
>     1. I run the project from the Godot Editor (as opposed to Run from inside VSCode)
> 
>     2. There is an error
> 
>     3. The Godot Editor opens the script for debugging (but doesn't switch to the script window automatically)
> 
>     4. The script remains open in the background
> 
>     5. Any changes I make to the script in VSCode are overwritten the next time I run the game from the Godot Editor b/c it is giving preference to the open script in the Godot Script Editor
> 
> 
> How I solved it:
> 
> Script Editor > Debug Menu > Debug With External Editor
> 
> Once I checked this box I no longer had the problem.

Thanks, I'll try that out, since I'm still experiencing this issue and just lost a bunch of code changes because I didn't notice... really painful :D 

As of now, this still happens to me in v4.3.stable.official [77dcf97d8], but as I understand from this issue history there will hopefully be a fix in 4.4?

**Comment by jasoben** [comment]
@christian-reichart just to clarify, you changed this setting (as seen in screenshot)? If so, it should not be opening the script in the Godot script editor at all, which is what causes the issue. (Unless there is another related bug I'm unaware of!)

<img width="320" alt="Image" src="https://github.com/user-attachments/assets/4b0cdb95-473d-47f7-8e1d-13f2374e409e" />

**Comment by christian-reichart** [comment]
> [@christian-reichart](https://github.com/christian-reichart) just to clarify, you changed this setting (as seen in screenshot)? If so, it should not be opening the script in the Godot script editor at all, which is what causes the issue. (Unless there is another related bug I'm unaware of!)
> <img alt="Image" width="320" src="https://private-user-images.githubusercontent.com/7291792/413504013-4b0cdb95-473d-47f7-8e1d-13f2374e409e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk2MDM2MzcsIm5iZiI6MTczOTYwMzMzNywicGF0aCI6Ii83MjkxNzkyLzQxMzUwNDAxMy00YjBjZGI5NS00NzNkLTQ3ZjctOGUxZC0xM2YyMzc0ZTQwOWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTVUMDcwODU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmYzNTE4ZjMzYmM2NmE3ZTFhNTMxY2MwYmFiMTIxNDZlMjQxZmQ2YjFmMmRjNzg3NGEyNGIyZTAyODgxZTFmNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GGYIZgrr5ZCmcw9pYeOOLu5xSVexyg6XOOhuiip2r2M">

Yes, with the setting enabled, the issue doesn't happen to me anymore. I just wasn't sure if the issue itself should have already been fixed in 4.3 or not.
